### PR TITLE
Not allow removing a single node with ref tables

### LINKED ISSUE #4104: Removing the only worker node with only reference table
I realized that if we have a 1 + 1 cluster (coordinator can't have placements) and create a table, say reference table then if we remove the worker node, we don't get any error, however if we try to add it back, we get the following error:

```
postgres=# select master_remove_node('localhost', 5001);
 master_remove_node
--------------------                                                                                                              (1 row)

postgres=# select master_add_node('localhost', 5001);
WARNING:  could not find any shard placements for shardId 102008
ERROR:  could not find any healthy placement for shard 102008
```

I would expect for `master_remove_node` to error since the coordinator doesn't have any placement for the reference table and we are removing the only worker.

### PR DESCRIPTION
**Pull request by SaitTalhaNisanci**
DESCRIPTION: Fix removal of a single node with only reference tables

We should not allow removing a node if it is the only node in the
cluster and there is a data on it. We have this check for distributed
tables but we didn't have it for reference tables.

Fixes #4104.

### CODE REVIEW

File: src/backend/distributed/metadata/node_metadata.c [hunk 1]
```diff
@@ -1054,14 +1055,15 @@ RemoveNodeFromCluster(char *nodeName, int32 nodePort)
 	WorkerNode *workerNode = ModifiableWorkerNode(nodeName, nodePort);
 	if (NodeIsPrimary(workerNode))
 	{
+		if (CanRemoveReferenceTablePlacements())
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1058]: I don't want to over  complicate things, but it'd be really nice if the error message is more clear when remove node fails because of reference tables

```C
 if (list_lenght(ReferenceTableOidList()) > 0)
{
if (CanRemoveReferenceTablePlacements())
{
    DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
}
else
{
    ereport (you cannot remove node because there are reference tables and  you would lose the data in reference tables 
     errhint drop all reference  tables before removing the node
}

}
```

And also optionally update the error message for `NodeGroupHasShardPlacements ` check stating that there are placements of  distributed tables

**Review by onderkalaci — COMMENTED** [review]
we might potentially break some user scripts where they used to be able to remove all nodes irrespective of existence of reference tables.

Still, this patch seems better alternative, users need to drop reference tables

File: src/backend/distributed/metadata/node_metadata.c (hunk 1 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/node_metadata.c:1058]: >you cannot remove node because there are reference tables and  you would lose the data in reference tables

What if there are reference tables and also distributed tables? Then this error message will first say that there are reference tables and if we drop the reference tables, it will error for distributed tables, however in the first error message it wasn't mentioning about the distributed tables. I guess that might be okay but I am curious about what you think

Comment (marcocitus) [reply src/backend/distributed/metadata/node_metadata.c:1058]: Seems ok to me.

Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/node_metadata.c:1057]: Added this block:

https://github.com/citusdata/citus/pull/4127/files#diff-b22dddbebb093ef1571e8253f2edd10aR1069-R1076

File: src/backend/distributed/metadata/node_metadata.c [hunk 2]
```diff
@@ -1053,17 +1054,26 @@ RemoveNodeFromCluster(char *nodeName, int32 nodePort)
...
-		DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
-
 		if (NodeGroupHasShardPlacements(workerNode->groupId,
 										onlyConsiderActivePlacements))
 		{
+			if (ClusterHasReferenceTable())
+			{
+				ereport(ERROR, (errmsg(
+									"you cannot remove node because there are reference "
+									"tables and you would lose the data on reference tables"),
```
Comment (onurctirtir) [inline src/backend/distributed/metadata/node_metadata.c:1073]: I think we do not prefer "you" word in error messages + emphasizing that it is the "last" node::

```suggestion
									"cannot remove the last worker node because there are reference "
									"tables and it would cause data loss for reference tables"),
```

File: src/test/regress/expected/single_node.out [hunk 3]
```diff
@@ -313,6 +313,11 @@ SELECT count(*) from local;
 
 ROLLBACK;
 RESET citus.task_executor_type;
+DROP TABLE test;
+-- cannot remove since a ref table exists on coordinator and no other nodes is added
```
Comment (onurctirtir) [inline src/test/regress/expected/single_node.out:317]:
```suggestion
-- cannot remove coordinator since a reference table exists on coordinator and no other worker nodes are added
```

File: src/backend/distributed/metadata/node_metadata.c [hunk 4]
```diff
@@ -1053,17 +1054,26 @@ RemoveNodeFromCluster(char *nodeName, int32 nodePort)
...
+								errhint(
+									"drop all reference tables before removing the node")));
+			}
 			ereport(ERROR, (errmsg("you cannot remove the primary node of a node group "
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1077]: We  have this question on slack right now. Can we now improve `you cannot remove the primary node of a node group which has shard placements` to have an hint to drop distributed tables?

Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/node_metadata.c:1077]: Actually should we mention `undistribute_table` here and also in reference table case? Because if people drop the table then they lose the data :)

File: src/backend/distributed/metadata/node_metadata.c (hunk 2 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/node_metadata.c:1073]: I tried to make it similar with the existent one. I guess we can change both of them or none of them.

File: src/backend/distributed/metadata/node_metadata.c [hunk 5]
```diff
@@ -1053,19 +1054,30 @@ RemoveNodeFromCluster(char *nodeName, int32 nodePort)
...
+									"you cannot remove node because there are reference "
+									"tables and you would lose the data on reference tables"),
+								errhint(
+									"use undistributed_table UDF to undistribute the reference tables")));
+			}
 			ereport(ERROR, (errmsg("you cannot remove the primary node of a node group "
-								   "which has shard placements")));
+								   "which has shard placements"),
+							errhint(
+								"use undistribute_table UDF to undistribute the distributed tables")));
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1080]: `use undistribute_table UDF to undistribute the distributed tables` we have a repetition that makes it slightly hard to follow.

So maybe `To proceed, either drop the distributed tables or use undistribute_table() function to convert them to local tables`  and  

`To proceed, either drop the reference tables or use undistribute_table() function to convert them to local tables`

**Review by onderkalaci — APPROVED** [review]
One minor improvement idea on the error message, other than that it looks good to me. (Also, please consider  Onur's comment  as well)

File: src/test/regress/sql/single_node.sql [hunk 6]
```diff
@@ -361,6 +361,9 @@ END;$$;
 SELECT * FROM pg_dist_node;
 SELECT create_distributed_function('call_delegation(int)', '$1', 'test');
 CALL call_delegation(1);
+DROP TABLE test CASCADE;
+-- cannot remove since a ref table exists on coordinator and no other nodes is added
```
Comment (SaitTalhaNisanci) [inline src/test/regress/sql/single_node.sql:365]:
```suggestion
-- cannot remove coordinator since a reference table exists on coordinator and no other worker nodes are added
```

**Review by onurctirtir — APPROVED** [review]
